Pairing API (#274)
WalletConnect/walletconnect-docs: docs/specs/pair/README.md

> # Pairing API Overview

Reviewer: Are we calling it `Pair` or `pairing`?

Directories say `pair`, but text talks about Pairing. Also `pair` is 4 letters so I'm sure @pedrouid has thoughts 😂

Author: yeah, hopefully we will resolve this issue on general sdk, today xd

## Description

Pairing API is a lightweight api for establishing an encrypted, protocol agnostic communication layer between peers. It's purpose is to provide a secure channel for proposing protocols or sending requests.

## Context

> WalletConnect offers Sign, Auth and Push Protocols. In order to allow reusable communitaction channel between peers pairing API exposes a common interface and allows for sending and receiving multi-protocol requests over a single pairing.

Reviewer: Typo:

```suggestion
WalletConnect offers Sign, Auth and Push Protocols. In order to allow reusable communication channel between peers, the Pairing API exposes a common interface and allows for sending and receiving multi-protocol requests over a single pairing.
```

This significantly improves end user experience by allowing to scan just a single QR code and receive requests from different protocols.

## User Flow

> User visits a new website that requires Sign session for submitting transactions Auth for authentication

Reviewer: ```suggestion
User visits a new website that requires a Sign session for submitting transactions, but uses Auth for authentication
```

1. Website displays qrcode or deep link
2. User scans qrcode or redirects to wallet

> 3. Popup with session proposal and authentication request appears
> 3. User approves session request and authenticates
> 4. User returns to website after prompt success
> 5. User is now authenticated and has a sign session established with the dapp

Reviewer 2: Nit: ordering
```suggestion
3. Popup with session proposal and authentication request appears
4. User approves session request and authenticates
5. User returns to website after prompt success
6. User is now authenticated and has a sign session established with the dapp
```

Also, I think we should somehow merge Auth request and Sign session proposal into one pop-up to have better UX

Author: that may be more on wallet to side to handle this.
keep in mind that requests will be received one by another, so if we do not send jsonrpc requests batched for how long we should wait to receive more requests? it may weaken UX as it won't be instant.
